### PR TITLE
Options/DeviceManager control button

### PR DESCRIPTION
**Pull request by mgamis-msft**
# What

- OptionsButton updated so that it works when using usePropsFor. 
- ControlBar and OptionsButtion stories re-using one example to avoid redundancy
- Temporarily deduplication devices from Calling SDK.

Clip of stories in storybook:
https://user-images.githubusercontent.com/79475487/120736107-816abe80-c4a0-11eb-988a-8e68789f1e93.mp4

# Why

https://skype.visualstudio.com/SPOOL/_sprints/taskboard/Web%20UI/SPOOL/CY2021-Q2/Sprint%2038%20(Ends%20June%2013)?workitem=2463874
https://skype.visualstudio.com/SPOOL/_sprints/taskboard/Web%20UI/SPOOL/CY2021-Q2/Sprint%2038%20(Ends%20June%2013)?workitem=2463880

# How Tested

I tested this by using usePropsFor('OptionsButton') and spreading the props onto OptionsButton in the sample Calling app.
I also tested the ControlBar and OptionsButton stories on my local storybook.

# Process & policy checklist

- [x] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/storybook/stories/Examples/TeamsInterop/__snapshots__/Lobby.stories.storyshot [hunk 1]
```diff
@@ -43,7 +43,7 @@ exports[`storybook snapshot tests Storyshots Examples/Teams Interop/Lobby Lobby
             style={
               Object {
                 "left": "50%",
-                "marginTop": "2.125rem",
+                "marginTop": "-2.125rem",
```
Comment (mgamis-msft) [inline packages/storybook/stories/Examples/TeamsInterop/__snapshots__/Lobby.stories.storyshot:46]: I noticed that this is supposed to be negative to pull the "You're in the lobby" div up to compensate for the ControlBar

File: packages/calling-stateful-client/src/DeviceManagerDeclarative.ts [hunk 2]
```diff
@@ -70,12 +70,12 @@ class ProxyDeviceManager implements ProxyHandler<DeviceManager> {
   };
 
   private videoDevicesUpdated = async (): Promise<void> => {
-    this._context.setDeviceManagerCameras(await this._deviceManager.getCameras());
+    this._context.setDeviceManagerCameras(dedupeVideoDevices(await this._deviceManager.getCameras()));
```
Comment (mgamis-msft) [inline packages/calling-stateful-client/src/DeviceManagerDeclarative.ts:73]: These dedupes are intended to be temporary because duplicates are being returned from the SDK currently

File: packages/calling-stateful-client/src/CallContext.ts [hunk 3]
```diff
@@ -571,7 +571,7 @@ export class CallContext {
   public setDeviceManagerCameras(cameras: VideoDeviceInfo[]): void {
     this.setState(
       produce(this._state, (draft: CallClientState) => {
-        if (!!draft.deviceManager.cameras && cameras.length > 0) {
+        if ((!draft.deviceManager.cameras || draft.deviceManager.cameras.length === 0) && cameras.length > 0) {
```
Comment (mgamis-msft) [inline packages/calling-stateful-client/src/CallContext.ts:574]: The logic is supposed to be 'only if we previously had no cameras and getting new cameras then we should select the first one by default'

File: packages/calling-stateful-client/src/DeviceManagerDeclarative.ts [hunk 4]
```diff
@@ -141,6 +141,32 @@ class ProxyDeviceManager implements ProxyHandler<DeviceManager> {
   }
 }
 
+/** Temporary helper function to dedupe duplicate video devices obtained from SDK */
+const dedupeVideoDevices = (devices: VideoDeviceInfo[]): VideoDeviceInfo[] => {
```
Comment (JamesBurnside) [inline packages/calling-stateful-client/src/DeviceManagerDeclarative.ts:145]: nit: use generics to avoid duplicate code, e.g.
```typescript
const dedupe = <T extends VideoDeviceInfo | AudioDeviceInfo>(devices: T[]): T[] => {
  const ids = new Set();
  const uniqueDevices: T[] = [];
  devices.forEach((device: T) => {
    if (!ids.has(device.id)) {
      uniqueDevices.push(device);
      ids.add(device.id);
    }
  });
  return uniqueDevices;
};
```

File: packages/react-components/src/components/OptionsButton.tsx [hunk 5]
```diff
@@ -63,13 +64,18 @@ const generateDefaultMenuProps = (props: OptionsButtonProps): { items: Array<any
           title: val.name,
           canCheck: true,
           isChecked: val.id === selectedCamera?.id,
-          onClick: () => !(val.id === selectedCamera?.id) && onSelectCamera(val)
+          onClick: () => {
+            if (val.id !== selectedSpeaker?.id) {
```
Comment (allenplusplus) [inline packages/react-components/src/components/OptionsButton.tsx:68]: Should it be selectedCamera here and not selectedSpeaker?

File: packages/react-components/src/components/OptionsButton.tsx [hunk 6]
```diff
@@ -81,13 +87,18 @@ const generateDefaultMenuProps = (props: OptionsButtonProps): { items: Array<any
           title: val.name,
           canCheck: true,
           isChecked: val.id === selectedMicrophone?.id,
-          onClick: () => !(val.id === selectedMicrophone?.id) && onSelectMicrophone(val)
+          onClick: () => {
+            if (val.id !== selectedSpeaker?.id) {
```
Comment (allenplusplus) [inline packages/react-components/src/components/OptionsButton.tsx:91]: Should this be selectedMicrophone here and not selectedSpeaker?

File: packages/calling-stateful-client/src/DeviceManagerDeclarative.ts (hunk 2 above)
Comment (prprabhu-ms) [reply packages/calling-stateful-client/src/DeviceManagerDeclarative.ts:73]: Please add a comment in the code with a TODO(...) referencing the bug / task against SDK that will remove the duplicates.

Comment (prprabhu-ms) [reply packages/calling-stateful-client/src/DeviceManagerDeclarative.ts:73]: Actually, better just add the bug reference near the definition of `dedupeAudioDevices` on line 145

File: packages/react-components/src/components/OptionsButton.tsx [hunk 7]
```diff
@@ -19,15 +19,15 @@ export interface OptionsButtonProps extends IButtonProps {
...
+  cameras?: { id: string; name: string }[];
   selectedMicrophone?: { id: string; name: string };
   selectedSpeaker?: { id: string; name: string };
   selectedCamera?: { id: string; name: string };
-  onSelectCamera?: (device: any) => Promise<void>;
-  onSelectMicrophone?: (device: any) => Promise<void>;
-  onSelectSpeaker?: (device: any) => Promise<void>;
+  onSelectCamera?: (device: any) => void;
+  onSelectMicrophone?: (device: any) => void;
+  onSelectSpeaker?: (device: any) => void;
```
Comment (allenplusplus) [inline packages/react-components/src/components/OptionsButton.tsx:30]: What is the reason for this change? Is it to fix some compiler issues? I believe the handlers for onSelectCamera, onSelectMicrophone, and onSelectSpeaker return Promise<void> , I'm wondering if we can keep it the same.

File: packages/calling-stateful-client/src/DeviceManagerDeclarative.ts (hunk 4 above)
Comment (prprabhu-ms) [reply packages/calling-stateful-client/src/DeviceManagerDeclarative.ts:145]: +1
I'd go further and define an anonymous interface that tells me exactly what I need to use the generic function:

```ts
const dedupe = <T extends {id: string}>(devices: T[]): T[] => {
  const ids = new Set();
  const uniqueDevices: T[] = [];
  devices.forEach((device: T) => {
    if (!ids.has(device.id)) {
      uniqueDevices.push(device);
      ids.add(device.id);
    }
  });
  return uniqueDevices;
};
```

File: packages/react-components/src/components/OptionsButton.tsx [hunk 8]
```diff
@@ -49,9 +49,10 @@ const generateDefaultMenuProps = (props: OptionsButtonProps): { items: Array<any
     onSelectSpeaker
   } = props;
 
-  const defaultMenuProps: { items: Array<any> } = { items: [] };
+  let defaultMenuProps: { items: Array<IContextualMenuItem> } | undefined = undefined;
```
Comment (prprabhu-ms) [inline packages/react-components/src/components/OptionsButton.tsx:52]: It's cleaner to let this be const and check later when using:

```ts
  const defaultMenuProps: { items: Array<IContextualMenuItem> } = { items: [] };

  ...

  if (defaultMenuProps.items.length === 0) {
    // Avoid creating an empty menu
    return undefined
  }
  return defaultMenuProps;
```

**Review by prprabhu-ms — APPROVED** [review]

File: packages/react-components/src/components/OptionsButton.tsx (hunk 7 above)
Comment (mgamis-msft) [reply packages/react-components/src/components/OptionsButton.tsx]: Yeah no problem. No compiler issues with keeping it the same. Interestingly, usePropsFor will still provide the handlers with the type (device: any) => void. I just thought a return type of void is less specific and allow more flexibility to use OptionsButton. But maybe we intend to use OptionsButton with our selectors.

File: packages/react-components/src/components/OptionsButton.tsx (hunk 5 above)
Comment (mgamis-msft) [reply packages/react-components/src/components/OptionsButton.tsx]: Another C/P error.. Thanks for spotting that.

File: packages/calling-stateful-client/src/DeviceManagerDeclarative.ts (hunk 4 above)
Comment (mgamis-msft) [reply packages/calling-stateful-client/src/DeviceManagerDeclarative.ts]: Thanks guys! I was trying to use generics to dedupe my dedupers by looking online but did not quite understand. Your suggestions really helped :)

File: packages/calling-stateful-client/src/DeviceManagerDeclarative.ts (hunk 2 above)
Comment (mgamis-msft) [reply packages/calling-stateful-client/src/DeviceManagerDeclarative.ts]: Unfortunately, I still haven't got confirmation from Calling that this is a bug and they have not logged a ticket. But I did add TODO to the deduper function for future reference.

**Review by allenplusplus — APPROVED** [review]

File: packages/react-components/src/components/OptionsButton.tsx [hunk 9]
```diff
@@ -36,7 +36,7 @@ export interface OptionsButtonProps extends IButtonProps {
  * @param props OptionsButtonProps
  * @returns MenuProps
  */
-const generateDefaultMenuProps = (props: OptionsButtonProps): { items: Array<any> } => {
+const generateDefaultMenuProps = (props: OptionsButtonProps): { items: Array<IContextualMenuItem> } | undefined => {
```
[the remark this answers is not in the document]
Comment (mgamis-msft) [reply packages/react-components/src/components/OptionsButton.tsx]: Good point. We don't really use the Array keyword. Changed for consistency.

File: packages/react-components/review/react-components.api.md [hunk 10]
```diff
@@ -221,20 +221,20 @@ export const OptionsButton: (props: OptionsButtonProps) => JSX.Element;
 // @public
 export interface OptionsButtonProps extends IButtonProps {
     // (undocumented)
-    cameras?: [{
+    cameras?: {
         id: string;
```
[the remark this answers is not in the document]
Comment (mgamis-msft) [reply packages/react-components/review/react-components.api.md]: Sure. Added OptionsDevice interface.
